py/mattak/backends/uproot/dataset.py

```diff
@@ -244,7 +255,67 @@ def eventInfo(self) -> Union[Optional[mattak.Dataset.EventInfo],Sequence[Optiona
     
     def N(self) -> int: 
         return self._hds.num_entries
+    
+    def __unpack_cal_parameters(self, cal_file : uproot.ReadOnlyDirectory) -> numpy.ndarray:
+        """
+        Function that reads out 9th order polynomial parameters from the calibration file
+
+        Returns
+        ---------
+        coef : np.ndarray of shape (24 * 4096, 10)
+        """
+        coef = numpy.stack(cal_file["coeffs_tree/coeff"].array(library = 'np'))   #stack is needed to convert ndarray of ndarrays to 'normally shaped' array, otherwise you have nested ndarrays
+        return coef
+
+    def __unpack_cal_residuals(self, cal_file : uproot.ReadOnlyDirectory) -> numpy.ndarray:
+        """
+        Function that reads out the residuals from the calibration file
+
+        Returns
+        ---------
+        (v_residuals, residuals) : both v_residuals and residuals have shape (points, 2)
+
+        """
+
+        vres_dac1, vres_dac2 = cal_file["aveResid_dac1"].values(axis = 0), cal_file["aveResid_dac2"].values(axis = 0)
+        residual_dac1, residual_dac2 = cal_file["aveResid_dac1"].values(axis = 1), cal_file["aveResid_dac2"].values(axis = 1)
+        return numpy.stack(numpy.array([vres_dac1, vres_dac2]), axis = -1), numpy.stack(numpy.array([residual_dac1, residual_dac2]), axis = -1)
+    
+    def __calibrate(self, waveform_array : numpy.ndarray, param : numpy.ndarray, 
+                    vres : numpy.ndarray, res : numpy.ndarray, starting_window : float | int,
+                    fit_min : float = -1.3, fit_max : float = 0.7, accuracy : float = 0.01) -> numpy.ndarray:
+        """
+        The calibration function that transforms waveforms from ADC to voltage
+
+        Parameters
+        ------------
+        waveform_array: array of one waveform, expected to have shape (24, 2048)
+        param : the parameters found in the calibration file, expected to have shape (24 * 4096, 10)
+        vres : the voltage points of the residuals
+        res : the ADC values of the residuals
+        starting_window : the sample on which the run started
+        
+        """
 
+        # "discrete" inverse
+        vsamples = numpy.arange(fit_min, fit_max, accuracy)
+        waveform_volt = numpy.zeros((24, 2048))
+        for c, wf_channel in enumerate(waveform_array):
+            starting_window_channel = starting_window[c]
+            # Reordering the parameters to match the correct starting window
+            # (calibration is done per sample)
+            param_channel = param[4096 * c : 4096 * (c + 1)]
```

Comment:
oh the data is stored in `(n_samples * n_channels)` ?

Reply:
Yes the parameters are stored as (24*4096, 10) as far as I can tell